srml/contract/src/rent.rs

```diff
@@ -0,0 +1,138 @@
+
+	if fee_per_block.is_zero() {
+		// The rent deposit offset reduced the fee to 0. This means that the contract
+		// gets the rent for free.
+		return false;
+	}
+
+	let blocks_to_rent = block_number.saturating_sub(contract.deduct_block);
+	let rent = fee_per_block
+		.checked_mul(&<BalanceOf<T>>::sa(blocks_to_rent.as_()))
+		.unwrap_or(<BalanceOf<T>>::max_value());
+	let subsistence_threshold = T::Currency::minimum_balance() + <Module<T>>::tombstone_deposit();
+
+	let rent_limited = rent.min(contract.rent_allowance);
+
+	let rent_allowance_exceeded = rent > contract.rent_allowance;
+	let is_below_subsistence = balance < subsistence_threshold;
+	let go_below_subsistence = balance.saturating_sub(rent_limited) < subsistence_threshold;
+	let can_withdraw_rent = T::Currency::ensure_can_withdraw(
+		account,
+		rent_limited,
+		WithdrawReason::Fee,
+		balance.saturating_sub(rent_limited),
+	).is_ok();
+
+	if !rent_allowance_exceeded && can_withdraw_rent && !go_below_subsistence {
+		// Collect dues
+
+		if pay_rent {
+			let imbalance = T::Currency::withdraw(
+				account,
+				rent,
+				WithdrawReason::Fee,
+				ExistenceRequirement::KeepAlive,
+			)
+			.expect(
+				"Withdraw has been checked above;
+						go_below_subsistence is false and subsistence > existencial_deposit;
+						qed",
+			);
+
+			<ContractInfoOf<T>>::mutate(account, |contract| {
+				contract
+					.as_mut()
+					.and_then(|c| c.as_alive_mut())
+					.expect("Dead or inexistent account has been exempt above; qed")
+					.rent_allowance -= imbalance.peek(); // rent_allowance is not exceeded
+			})
+		}
+		return false;
+	} else {
+		// Evict
+
+		if can_withdraw_rent && !go_below_subsistence {
+			T::Currency::withdraw(
+				account,
+				rent,
+				WithdrawReason::Fee,
+				ExistenceRequirement::KeepAlive,
+			)
+			.expect("Can withdraw and don't go below subsistence");
+		} else if !is_below_subsistence {
+			T::Currency::make_free_balance_be(account, subsistence_threshold);
+		} else {
+			T::Currency::make_free_balance_be(account, <BalanceOf<T>>::zero());
+		}
+
+		if !is_below_subsistence {
+			// Note: this operation is heavy.
+			// Note: if the storage hasn't been touched then it returns None, which is OK.
+			let child_storage_root = runtime_io::child_storage_root(&contract.trie_id);
+
+			let tombstone = TombstoneContractInfo::new(
+				child_storage_root,
+				contract.storage_size,
+				contract.code_hash,
+			);
+			<ContractInfoOf<T>>::insert(account, ContractInfo::Tombstone(tombstone));
+			runtime_io::kill_child_storage(&contract.trie_id);
+		}
+
+		return true;
+	}
+}
+
+/// Make account paying the rent for the current block number
+pub fn pay_rent<T: Trait>(account: &T::AccountId) {
+	try_evict_or_and_pay_rent::<T>(account, <system::Module<T>>::block_number(), true);
+}
+
+/// Make account paying the rent for the current block number
+pub fn try_evict_at<T: Trait>(account: &T::AccountId, block: T::BlockNumber) -> bool {
```

Comment:
Hmm, from the comment it is actually kind of hard to tell what the difference between these two functions

Reply:
updated